### PR TITLE
feat(lint): Verify Positional Format Strings

### LINKED ISSUE #10347: Add lint: "Multiple substitutions specified in non-positional format"
Nonfatal warnings:
```
> Task :AnkiDroid:mergePlayDebugResources
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values-hi/04-network.xml:103:4: Multiple substitutions specified in non-positional format of string resource string/sync_media_partial. Did you mean to add the formatted="false" attribute?
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values/04-network.xml:87:4: Multiple substitutions specified in non-positional format of string resource string/sync_media_partial. Did you mean to add the formatted="false" attribute?
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/res/values-hr/04-network.xml:103:4: Multiple substitutions specified in non-positional format of string resource string/sync_media_partial. Did you mean to add the formatted="false" attribute?
```

_Originally posted by @david-allison in https://github.com/ankidroid/Anki-Android/issues/10333#issuecomment-1039157947_

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
AAPT displayed these issues as build as warnings, but they did not break CI

## Fixes
Fixes #10347

## Approach

So, add some lint to make sure this occurs.

We take AAPT's C++ code and convert it to Kotlin.

Errors looks like:

```
AnkiDroid\src\main\res\values\04-network.xml:84:
Multiple substitutions specified in non-positional format.
Convert "%s" to "%1$s"
[NonPositionalFormatSubstitutions from com.ichi2.anki.lint]
```

## How Has This Been Tested?
Unit tested

Tested on commit d39249805081d18d1cf05178874608d161ae5a7a (failed)

## Learning (optional, can help others)
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/tools/aapt2/util/Util.cpp;drc=fbb7fe0da32896a871dd395845f1f510b075f8d5

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: lint-rules/src/main/java/com/ichi2/anki/lint/utils/Aapt2Util.kt [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            }
+
+            if (index < str.length) {
+                index++
+            }
+        }
+
+        if (argCount > 1 && nonpositional) {
+            // Multiple arguments were specified, but some or all were non positional.
+            // Translated
```
Comment (Arthur-Milchior) [inline lint-rules/src/main/java/com/ichi2/anki/lint/utils/Aapt2Util.kt:131]: no new line here

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/NonPositionalFormatSubstitutionsTest.kt [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    @Language("XML")
+    private val unambiguous = "<resources><string name=\"hello\">%1\$s %2\$s</string></resources>"
+
+    /** %% is an encoded percentage */
+    @Language("XML")
+    private val encoded = "<resources><string name=\"hello\">%%</string></resources>"
+
+    @Test
+    fun errors_if_ambiguous() {
+        // This appears to be a bug in StringCasingDetector - string is self-referential.
```
Comment (Arthur-Milchior) [inline lint-rules/src/test/java/com/ichi2/anki/lint/rules/NonPositionalFormatSubstitutionsTest.kt:48]: What do you mean by self-referential?

File: lint-rules/src/main/java/com/ichi2/anki/lint/utils/Aapt2Util.kt [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            // strings may rearrange the order of the arguments, which will break the
+            // string.
+            return false
+        }
+        return true
+    }
+
+    private fun consumeDigits(s: String, index: Int): Int {
```
Comment (Arthur-Milchior) [inline lint-rules/src/main/java/com/ichi2/anki/lint/utils/Aapt2Util.kt:139]:
```java
/**
* @return Number of digits in `s`, starting at position `index`.
*/
```

**Review by Arthur-Milchior — APPROVED** [review]
LGTM.

I have not checked in detail the algorithm itself, since it's a port anyway (and that it would take time to know the formating rules in much detail). Tests make sens, and in the worst case, if we ever get a false warning, it's okay to correct later

File: lint-rules/src/test/java/com/ichi2/anki/lint/rules/NonPositionalFormatSubstitutionsTest.kt (hunk 2 above)
Comment (david-allison) [reply lint-rules/src/test/java/com/ichi2/anki/lint/rules/NonPositionalFormatSubstitutionsTest.kt:48]: Good catch - copy and paste error

**Review by Arthur-Milchior — APPROVED** [review]

**Review by mikehardy — APPROVED** [review]
This looks good - just to verify I won't break master with the merge I pulled it locally and ran the CI suite after rebasing to master, all seemed fine

**Pull request merged**